### DISCUSSION TITLE
Building extension with fortran

### DISCUSSION THREAD

**Discussion by AntoineD** [pypa/cibuildwheel#764] [Q&A]
Hello,

in order to build an extension with fortran code, would it be possible to install a fortran compiler?
May be with `CIBW_BEFORE_BUILD_MACOS`, `CIBW_BEFORE_BUILD_WINDOWS` and `CIBW_BEFORE_BUILD_LINUX`?

**Comment by joerick** [comment]
Should be possible, yes. To install stuff on the system before build, you probably want [`CIBW_BEFORE_ALL`](https://cibuildwheel.readthedocs.io/en/stable/options/#before-all).

**Comment by AntoineD** [comment]
@joerick Do you know where I can find information about how to install stuff for each platform?

**Comment by joerick** [comment]
I do not know about Fortran compilers, sorry. But I do note that [scikit-build](https://scikit-build.readthedocs.io/en/latest/) has Fortran support, that might be worth a look!

**Reply by AntoineD** [reply]
Thanks @joerick, unfortunately it seems that scikit-build does not install compilers.
I'm just looking for a generic way to install stuff on each platforms, whatever stuff.
I'll handle later on for specifically installing a fortran compiler.

**Reply by joerick** [reply]
I see. Well, typically people use homebrew on Mac, chocolatey on Windows, and yum/apt on manylinux (depending on the manylinux image you're using)

**Reply by henryiii** [reply]
Manylinux (1-2014) uses the RH dev toolkit, so you should have access to that in yum. 2_24 has normal apt.

**Reply by AntoineD** [reply]
Do you know if homebrew and chocolatey are already bootstrapped?

**Reply by joerick** [reply]
That's up to the CI provider you're using, but I expect most preinstall homebrew and chocolately. [Github Actions and Azure Pipelines certainly do](https://github.com/actions/virtual-environments).

**Reply by AntoineD** [reply]
Thanks!

**Reply by mayeut** [reply]
manylinux images already provide a fortran compiler:
```
Matt % docker run -it --rm quay.io/pypa/manylinux1_x86_64 gfortran --version   
GNU Fortran (GCC) 4.8.2 20140120 (Red Hat 4.8.2-15)
Copyright (C) 2013 Free Software Foundation, Inc.

GNU Fortran comes with NO WARRANTY, to the extent permitted by law.
You may redistribute copies of GNU Fortran
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING

Matt % docker run -it --rm quay.io/pypa/manylinux2010_x86_64 gfortran --version 
GNU Fortran (GCC) 8.3.1 20190311 (Red Hat 8.3.1-3)
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Matt % docker run -it --rm quay.io/pypa/manylinux2014_x86_64 gfortran --version 
GNU Fortran (GCC) 9.3.1 20200408 (Red Hat 9.3.1-2)
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Matt % docker run -it --rm quay.io/pypa/manylinux_2_24_x86_64 gfortran --version
GNU Fortran (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

For windows & macOS, you might have some luck looking at how NumPy is doing that to build OpenBLAS:
https://github.com/MacPython/openblas-libs

**Reply by AntoineD** [reply]
Thanks @mayeut and all, I got almost all my wheels built!

**Reply by Nicholaswogan** [reply]
I'm using cibuildwheel with github actions to build wheels that depend on fortran code ([see here](https://github.com/Nicholaswogan/NumbaMinpack/runs/4362080106?check_suite_focus=true)). Linux and Windows builds just fine, but MacOS does not. I tried using `CIBW_BEFORE_ALL_MACOS: brew install gcc@11`, and the virtual machine claims it already has gcc installed (which includes gfortran). But cmake complains it can't find a fortran compiler: 

```
-- The Fortran compiler identification is unknown
    CMake Error at CMakeLists.txt:2 (project):
      No CMAKE_Fortran_COMPILER could be found.
    
      Tell CMake where to find the compiler by setting either the environment
      variable "FC" or the CMake cache entry CMAKE_Fortran_COMPILER to the full
      path to the compiler, or to the compiler name if it is in the PATH.
    
    
    -- Configuring incomplete, errors occurred!
```

Do you know how I might fix this problem?

**Reply by AntoineD** [reply]
@Nicholaswogan See [here](https://github.com/AntoineD/pdfo/blob/build-wheels/.github/workflows/build_wheels.yml) for a working workflow.
